Add custom IPC directory selector to the tournament client

MiraiSubject opened the pull request:
This changes the Refresh button in the `SetupScreen`  of the tournament client to have the functionality of manually specifying a CE Stable location.

The auto detection option has been moved to the directory selector and error handling has also been added for invalid locations or failure for autodetecting. 

This feature could be used in scenarios where users have multiple stable CE installations to separate the different tournaments. 

When the Tournament Client first boots up, it does the ENV_VAR, registry and LocalAppData check as usual, but saves the ipc path now to `tournament/stable.json`.

When a user forces auto detection in the directory selector the stablePath gets set to an empty string so ENV_VAR, registry and LocalAppData check can take place again. 

**Screenshots:**
![dotnet_2020-05-17_22-58-51](https://user-images.githubusercontent.com/803255/82162013-9da1ba80-98a1-11ea-8760-16d0e6643b6d.png)

![dotnet_2020-05-17_22-59-05](https://user-images.githubusercontent.com/803255/82162018-a72b2280-98a1-11ea-9cdf-fcb9294a823d.png)

![dotnet_2020-05-17_23-00-07](https://user-images.githubusercontent.com/803255/82162023-ad210380-98a1-11ea-9540-6ba9cb679b7c.png)

**GIF of the functionality:**
![2020-05-18_00-58-33](https://user-images.githubusercontent.com/803255/82162181-d2fad800-98a2-11ea-8a26-7b896d45a766.gif)

MiraiSubject committed:
Introduce model to store path of stable osu!

MiraiSubject committed:
Refactored stable path finding and added json config detection.

This also migrates the values found in the other methods to the configuration file.

MiraiSubject committed:
Dependency cache the ipc location file

MiraiSubject committed:
User interface setup for custom IPC location

Right now makes use of another ActionableInfo field. Probably a better idea to add an extra button to the Current IPC Storage actionable field.

MiraiSubject committed:
Merge remote-tracking branch 'upstream/master' into custom-ipc-location

MiraiSubject committed:
Update xmldoc with accurate information about the model

MiraiSubject committed:
Force a read of the location file during detection

MiraiSubject committed:
Moved refresh button to directoryselector

MiraiSubject committed:
Changed behaviour of refresh button in SetupScreen

MiraiSubject committed:
Make IPC error dialog reusable

and inspectcode fixes

MiraiSubject committed:
Move stablestorage check to path selection screen

Also forced stablepath to be empty during auto detection so it checks other sources to load ipc from

MiraiSubject committed:
Renamed Refresh button to Change source

MiraiSubject committed:
Fix test crashing because of sceneManager not being nullable

bdach commented on osu.Game.Tournament/TournamentGameBase.cs:
```diff
@@ -33,6 +33,8 @@ public abstract class TournamentGameBase : OsuGameBase
     {
         private const string bracket_filename = "bracket.json";
 
+        private const string stable_config = "tournament/stable.json";
```
Shouldn't be defined four times (here, in `FileBasedIPC`, in `StablePathSelectScreen` and in `SetupScreen`).

bdach commented on osu.Game.Tournament/IPC/FileBasedIPC.cs:
```diff
@@ -146,54 +156,143 @@ public Storage LocateStableStorage()
-                {
-                    using (RegistryKey key = Registry.ClassesRoot.OpenSubKey("osu"))
-                        stableInstallPath = key?.OpenSubKey(@"shell\open\command")?.GetValue(string.Empty).ToString().Split('"')[1].Replace("osu!.exe", "");
+                return stableInstallPath;
```
Nit, but I'd probably explicitly return null/empty string here for clarity (right now it looks a little weird, as if it somehow found a path after trying everything and failing).

bdach commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -0,0 +1,203 @@
+            var target = directorySelector.CurrentDirectory.Value.FullName;
+            Logger.Log($"Changing Stable CE location to {target}");
+
+            if (File.Exists(Path.Combine(target, "ipc.txt")))
```
I'd invert and deal with the negative part first (`ipc.txt` not found), and early-return if it didn't succeed.

This is also a piece of code (albeit a one-liner) that has duplicated filenames (see `FileBasedIPC.checkExists`). Would like to see at least the filename extracted to a common constant.

bdach commented on osu.Game.Tournament/IPC/FileBasedIPC.cs:
```diff
@@ -146,54 +156,143 @@ public Storage LocateStableStorage()
+
+        private string findFromJsonConfig()
+        {
+            try
+            {
+                Logger.Log("Trying to find stable through the json config");
+
+                if (!string.IsNullOrEmpty(stableInfo.StablePath.Value))
+                    return stableInfo.StablePath.Value;
             }
-            finally
+            catch
             {
-                Logger.Log($"Stable path for tourney usage: {stableInstallPath}");
             }
+
+            return null;
+        }
```
This method has an actively misleading name (it doesn't read any JSON config or find anything in one, it just reads a previously-found path).

bdach commented on osu.Game.Tournament/IPC/FileBasedIPC.cs:
```diff
@@ -146,54 +156,143 @@ public Storage LocateStableStorage()
+
+            if (checkExists(stableInstallPath))
+            {
+                stableInfo.StablePath.Value = stableInstallPath;
+                saveStablePath();
+                return stableInstallPath;
```
This scheme of writing to the cached `StableInfo` and then returning the local is very strange. Why not just return `stableInstallPath`, and then assign to `StablePath.Value` and call `saveStablePath` higher up in `findStablePath`?

(Better yet, you could probably assign to the bindable inside of `saveStablePath`, too.)

bdach commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -0,0 +1,203 @@
+
+        private void autoDetect()
+        {
+            stableInfo.StablePath.Value = string.Empty; // This forces findStablePath() to look elsewhere.
```
`LocateStableStorage` is the only caller of `findStablePath` - so why not just remove the misleading `findFromJsonConfig`, which would render this unnecessary?

bdach commented on osu.Game.Tournament/TournamentGameBase.cs:
```diff
@@ -141,6 +145,23 @@ private void load(Storage storage, FrameworkConfigManager frameworkConfig)
             });
         }
 
+        private void readStableConfig()
+        {
+            if (storage.Exists(stable_config))
+            {
+                using (Stream stream = storage.GetStream(stable_config, FileAccess.Read, FileMode.Open))
+                using (var sr = new StreamReader(stream))
+                {
+                    stableInfo = JsonConvert.DeserializeObject<StableInfo>(sr.ReadToEnd());
+                }
+            }
+
+            if (stableInfo == null)
+                stableInfo = new StableInfo();
+
+            dependencies.Cache(stableInfo);
+        }
```
Not sure I like the fact that the json is being read here, but written in `FileBasedIPC`.

MiraiSubject committed:
Use one constant for STABLE_CONFIG location string

MiraiSubject committed:
Return null if path is not found, for clarity

MiraiSubject committed:
invert if-statement and early return + reuse of checkExists

MiraiSubject committed:
use common const in TournamentGameBase

MiraiSubject committed:
removed redundant code and use existing checkExists

MiraiSubject committed:
make CheckExists static public and removed unnecessary code

MiraiSubject committed:
Merge branch 'master' into custom-ipc-location

MiraiSubject committed:
Add StablePathSelectScreen visual test

MiraiSubject committed:
Merge branch 'master' into custom-ipc-location

MiraiSubject committed:
corrected styling issues

bdach commented on osu.Game.Tournament/TournamentGameBase.cs:
```diff
@@ -141,6 +143,23 @@ private void load(Storage storage, FrameworkConfigManager frameworkConfig)
             });
         }
 
+        private void readStableConfig()
```
I still don't quite understand why this has to be here. Can this not be done inside of `FileBasedIPC`?

MiraiSubject replied:
The reason it's being read here is because it's being it's being put in the dependency container after it was read.
`dependencies.Cache(stableInfo);` 

 Moving the read to `FileBasedIPC` and losing DI on it means that `SetupScreen` can't use this binding any more to refresh the values on screen:

```
stableInfo.StablePath.BindValueChanged(_ =>
{
    Schedule(reload);
});
```

and `StablePathSelectScreen` then needs to access the value of `stableInfo` through `FileBasedIPC`

bdach commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -0,0 +1,177 @@
+        [BackgroundDependencyLoader(true)]
+        private void load(Storage storage, OsuColour colours)
+        {
+            // begin selection in the parent directory of the current storage location
+            var initialPath = new DirectoryInfo(storage.GetFullPath(string.Empty)).Parent?.FullName;
+
+            if (!string.IsNullOrEmpty(stableInfo.StablePath.Value))
+            {
+                // If the original path info for osu! stable is not empty, set it to the parent directory of that location
+                initialPath = new DirectoryInfo(stableInfo.StablePath.Value).Parent?.FullName;
+            }
```
Can be simplified to

```csharp
var initialPath = new DirectoryInfo(storage.GetFullPath(stableInfo.StablePath.Value ?? string.Empty)).Parent?.FullName;
```

bdach commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -0,0 +1,177 @@
+                AddInternal(overlay);
+                Logger.Log("Folder is not an osu! stable CE directory");
+                return;
+                // Return an error in the picker that the directory does not contain ipc.txt
```
This comment reads as if there is something missing here, which I don't think is the case (there's no error being returned). I'd just remove it.

bdach commented on osu.Game.Tournament.Tests/Screens/TestSceneStablePathSelectScreens.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace osu.Game.Tournament.Tests.Screens
+{
+    public class TestSceneStablePathSelectScreens : TournamentTestScene
```
```suggestion
    public class TestSceneStablePathSelectScreen : TournamentTestScene
```
(+ ctor)

bdach commented on osu.Game.Tournament/Models/StableInfo.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        public Bindable<string> StablePath = new Bindable<string>(string.Empty);
+
+        [JsonIgnore]
+        public const string STABLE_CONFIG = "tournament/stable.json";
```
I would prefer this to also reside in `FileBasedIPC` if the read can be moved there.

bdach commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -0,0 +1,177 @@
+
+            if (fileBasedIpc?.IPCStorage == null)
+            {
+                // Could not auto detect
```
This one is also not really necesssary.

MiraiSubject committed:
Remove unnecessary comments, simplify initialPath and clarified TestScene name

MiraiSubject committed:
move stable config declaration and initial reading

MiraiSubject committed:
Merge branch 'master' into custom-ipc-location

MiraiSubject commented:
I sloppily pushed some code with errors in it. VSCode show me these errors till just now. I'm right on it

MiraiSubject committed:
Move stable_config declaration, rename testscene

MiraiSubject committed:
Merge branch 'master' into custom-ipc-location

MiraiSubject committed:
Merge branch 'master' into custom-ipc-location

MiraiSubject commented:
This PR is broken as of a recent update somewhere in the chain where a path gets read. I discovered this while moving the reads of `stableInfo` to `FileBasedIPC`. As a sanity check I stashed my changes and used my old implementation, but the same exception gets thrown now. 

This exception now gets thrown when calling `StablePathSelectScreen`:

```An unhandled exception of type 'System.ArgumentException' occurred in osu.Framework.dll: '"F:\SSDPrograms\osu! Tourney" traverses outside of "C:\Users\Subject\AppData\Roaming\osu" and is probably malformed'```

This exception initially points to this line in osu-framework: 
https://github.com/ppy/osu-framework/blob/8b00dbb68fb00b98db64644836dec8aadbec4eaa/osu.Framework/Platform/NativeStorage.cs#L65

But this check was already there when I initially created the PR, so I have no idea what else might have changed to trigger this exception. 

In the current state at least it's impossible to set a starting directory outside the original osu! lazer base path.

The specific line in my PR that breaks is: 

`var initialPath = new DirectoryInfo(storage.GetFullPath(stableInfo.StablePath.Value ?? string.Empty)).Parent?.FullName;`

Changing it to 
`var initialPath = new DirectoryInfo(storage.GetFullPath(string.Empty)).Parent?.FullName`

makes it work again, but the starting directory will then always be osu! lazer's base path, instead of the currently selected IPC location. 

I'll be implementing this change with the other requested changes regarding the read of `stableInfo` and await another review.

peppy commented:
If you need to create a new storage outside of the previous bounds, you may want to look at using this new method available from `GameHost` (you can use DI to resolve this):

https://github.com/ppy/osu-framework/blob/6083659ade1b81deb9e0e44f1ddb6367c8af0c43/osu.Framework/Platform/GameHost.cs#L132

MiraiSubject committed:
Moved stableInfo read to FileBasedIPC

DI is also not needed anymore to access StableInfo, this goes through FileBasedIPC.
Note: directory selector now always navigates to the osu! lazer base path.

MiraiSubject committed:
use GameHost's GetStorage instead of local storage

This will now get the IPC Path again as the default path if one is present, else it will fall back to osu! lazer's base path.

MiraiSubject committed:
Fix test crashing: NullReferenceException

MiraiSubject committed:
Fixed path is empty exception

Also converted method to property get, private set

MiraiSubject committed:
Merge branch 'master' into custom-ipc-location

bdach commented on osu.Game.Tournament/Components/IPCErrorDialog.cs:
```diff
@@ -0,0 +1,26 @@
+                new PopupDialogOkButton
+                {
+                    Text = @"Alright.",
+                    Action = () => { Expire(); }
```
Braces are unnecessary.
```suggestion
                    Action = () => Expire()
```

bdach commented on osu.Game.Tournament/IPC/FileBasedIPC.cs:
```diff
@@ -146,54 +155,135 @@ public Storage LocateStableStorage()
+            return IPCStorage;
         }
 
+        public static bool CheckExists(string p) => File.Exists(Path.Combine(p, "ipc.txt"));
```
I'd rename to something like `IPCFileExistsInDirectory`. This method should ideally not be `public`, too.

bdach commented on osu.Game.Tournament/IPC/FileBasedIPC.cs:
```diff
@@ -146,54 +155,135 @@ public Storage LocateStableStorage()
         {
-            static bool checkExists(string p) => File.Exists(Path.Combine(p, "ipc.txt"));
+            if (!string.IsNullOrEmpty(readStableConfig()))
+                return StableInfo.StablePath.Value;
```
I find the fact that this returns a path only in this one case, and `null` in all others, really strange. This API is unintuitive to use.

If I got a `null` path from a method called `findStablePath()`, I'd assume that a stable path wasn't found - which is not the case, it means that `FileBasedIPC` found the path on its own, which is really counterintuitive IMO.

MiraiSubject replied:
> If I got a null path from a method called findStablePath(), I'd assume that a stable path wasn't found - which is not the case, it means that FileBasedIPC found the path on its own, which is really counterintuitive IMO.

I was very surprised when I first read this, because it was intended to also return the path for all other methods.  I then noticed that I committed a null I used to test a failure case (namely CI failing). This was very sloppy on my part and I will revert that right away.

bdach commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -0,0 +1,173 @@
+            var fileBasedIpc = ipc as FileBasedIPC;
+            Logger.Log($"Changing Stable CE location to {target}");
+
+            if (!FileBasedIPC.CheckExists(target))
+            {
+                overlay = new DialogOverlay();
+                overlay.Push(new IPCErrorDialog("This is an invalid IPC Directory", "Select a directory that contains an osu! stable cutting edge installation and make sure it has an empty ipc.txt file in it."));
+                AddInternal(overlay);
+                Logger.Log("Folder is not an osu! stable CE directory");
+                return;
+            }
+
+            fileBasedIpc?.SaveStableConfig(target);
+            fileBasedIpc?.LocateStableStorage();
```
This is calling three things in sequence on `fileBasedIpc`, which means that this snippet of code is a good candidate for encapsulating in a method. Better yet, if you exposed a method in `FileBasedIPC` like:

```csharp
public bool SetIPCLocation(string path);
```

then you could probably (a) get rid of the strange `string` return type and config lookup case in `getStablePath()`, *and* (b) you could probably reuse *that same* method inside of the stable path lookup code.

bdach reviewed (dismissed):
I was going to try to fix this up myself, but reading through I still have objections to how this code is structured, so I'll leave a review instead.

@MiraiSubject If you're tired of applying suggestions, then I can do a refactoring pass over this, but I'd need time to get to it.

MiraiSubject committed:
Apply review suggestions

MiraiSubject commented:
I've applied your suggestions! All of them made sense and one of them was just a really sloppy error that I had slip through.

bdach commented:
I'm still not a fan of how `findStablePath` reads the json config. This makes it so that if you change the ipc location once manually, autodetect stops working as it'll just read its value over and over again.

I tried to make the change myself but it'd involve significant restructuring of `FileBasedIPC` so I'd rather hold off on that until #9210 is merged as that looks like a straight-forward fix and I'd like to avoid conflicts.

MiraiSubject commented:
Yeah I'm alright with that! I've been facing this problem myself but just haven't figured out a solution for it yet.

I had the idea of skipping the json file read when pressing auto detect so it tries to detect the other methods first, but that would indeed require significant restructuring. I believe I had this working in an older commit but it was very messy.

bdach committed:
Merge branch 'master' into custom-ipc-location

bdach committed:
Rework StableInfo into a DI'd data structure

bdach committed:
Remove unused argument

bdach committed:
Allow auto-detect to work after choosing manually

bdach committed:
Streamline autodetect & manual set path

bdach committed:
Fix silent failure

bdach committed:
Remove unnecessary members

bdach committed:
Add xmldoc

bdach commented:
I've made significant changes to this. I think it's tolerable now, but there might be more rough edges. Major changes include:

* factoring out `StableInfo` into a self-contained DI'd structure (which I think makes everything flow better),
* improving error-handling (return value of the old `locateStableStorage()` was being ignored),
* removing some unused arguments & members,
* privatising several members,
* adding xmldoc.

If you disagree with any of the changes let me know and I'll consider reverting.

bdach committed:
Check for nulls when looking for ipc.txt

MiraiSubject commented on osu.Game.Tournament/Screens/StablePathSelectScreen.cs:
```diff
@@ -42,11 +38,6 @@ private void load(Storage storage, OsuColour colours)
         {
             var initialPath = new DirectoryInfo(storage.GetFullPath(string.Empty)).Parent?.FullName;
 
-            if (!string.IsNullOrEmpty(stableInfo.StablePath))
```
This is a little nitpick, but right now the DirectorySelector always defaults to the lazer data directory. In many cases users have their stable tournament install located next to their normal installation in the form of: 

``` 
<ParentDirectory>/osu!
<ParentDirectory>/osu! Tourney
```

Having it default to the parent directory of the default install (if detected) would make it more easier for the user to the get to the correct installation.

bdach replied:
I was going to keep it but forgot about it along the way. The previous behaviour should be back now with my latest commit.

bdach committed:
Bring back initial directory behaviour

MiraiSubject reviewed (commented):
Was able to go through everything now and it looks good to me. API makes much more sense now.

I forget sometimes that it's just normal OOP and that there's nothing weird going on here, so I tend to restrict myself to what I already see.

peppy committed:
Merge branch 'master' into custom-ipc-location

peppy committed:
Simplify lookup fallback code

peppy committed:
try-catch registry lookup to avoid crashes on non-windows platforms

peppy commented:
Made some minor changes to this, but feels pretty great otherwise 👍 .
